### PR TITLE
Add safety comments

### PR DESCRIPTION
**Pull request by Manishearth**
I recently performed safety review on 0.19. Besides finding the already-fixed https://github.com/huggingface/tokenizers/issues/1491, I also found that these cases of unsafety were a bit hard to disentangle, and documented them further.

### CODE REVIEW

**Comment by sftse** [comment]
Does the referenced UCG discussion [412](https://github.com/rust-lang/unsafe-code-guidelines/issues/412) apply to this case? Exercising the offending code under miri with and without the flags mentioned in the issue does not make a difference.
```
MIRIFLAGS='-Zmiri-env-set=TOKENIZERS_PARALLELISM=false -Zmiri-recursive-validation' cargo +nightly miri test bpe::trainer::tests::test_train
```

**Comment by Manishearth** [comment]
@sftse IIRC miri doesn't currently do anything special around Vecs in this way, so it's not useful to rely on it for whether the rules apply. I did mention "this and related issues" because I know this type of thing has been the subject of discussion. For now I'm comfortable giving a sense of "this is potentially but not probably problematic"

**Comment by Manishearth** [comment]
Oh! The & to &mut transmute is the problem, I see. We can use `&raw` or `addr_of` there instead. The recursive validity issue made me not notice the other issue there.

Given that I might just write it with manual unsafe (which would not be open to recusrive validity issues or transmute issues) then.

**Comment by sftse** [comment]
Thanks!

Before seeing your PR I had made #1664, trying to hit the same angle but a bit more verbose, I should have mentioned that. #1651 and #1664 have a bit of overlap, the maintainers may want to deduplicate how they see fit.

File: tokenizers/src/models/bpe/trainer.rs [hunk 1]
```diff
@@ -530,14 +530,30 @@ impl BpeTrainer {
...
-                    // We can merge each of these words in parallel here because each position
-                    // can be there only once (HashSet). So this is safe.
+                    // Safety:
+                    // We are producing a valid pointer since we are indexing in bounds
+                    //
+                    // We can access each `word` here in parallel because each position
+                    // can be there only once (pos is a HashSet).
                     unsafe {
+                        assert!(i < words_len);
+                        // This is words[i], but avoids needing to go through &T (which triggers UB)
```
Comment (ArthurZucker) [inline tokenizers/src/models/bpe/trainer.rs:555]: FMI, why is `&T` UB?

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot both! 🤗 quite honored to see core rust people here 👀
LGTM!

**Comment by ArthurZucker** [comment]
`cargo fmt` should be enough for the CI!

**Comment by ArthurZucker** [comment]
```
Error:    --> src/models/bpe/trainer.rs:535:17
    |
535 |             let ref pos: HashSet<usize> = top.pos;
    |             ----^^^^^^^--------------------------- help: try: `let pos: &HashSet<usize> = &top.pos;`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#toplevel_ref_arg
    = note: `-D clippy::toplevel-ref-arg` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::toplevel_ref_arg)]`
```
seems to cause an issue with compilation, not sure if it's actionable on my side

File: tokenizers/src/models/bpe/trainer.rs (hunk 1 above)
Comment (Manishearth) [reply tokenizers/src/models/bpe/trainer.rs:554]: It's UB to mutate data that currently has an active `&T`. This persists through raw pointers in some ways (and not in others).

**Comment by ArthurZucker** [comment]
Thanks!

**Pull request merged**